Defend against empty responses

gbirke opened the pull request:
Catgraph should always return at least the original category id. In case
it doesn't, we check it and provide an error message.

For this change, the error variable needs to be module-global - bad style?

gbirke committed:
Defend against empty responses

Catgraph should always return at least the original category id. In case
it doesn't, we check it and privide an error message.

For this change, the error variable needs to be module-global - bad style?

WMDE-Fisch commented on DeepCat.js:
```diff
@@ -152,12 +156,18 @@
 			newSearchTermString = '';
 
+			if ( !responses[i]['result'] || responses[i]['result'].length == 0) {
+				errorMessages[0] = createErrorMessage( 'deepcat-error-unexpected-response', null );
```
Why not pushing the error message directly to `error`?

gbirke replied:
Because that would cause the same error message if the result was empty multiple times. This way, we get either an array with 0 or 1 elements regardless of the number of empty results.

WMDE-Fisch commented on DeepCat.js:
```diff
@@ -166,6 +176,8 @@
 			newSearchTerms[userParameters['searchTermNum']] = newSearchTermString;
 		}
 
+		errors.concat(errorMessages);
```
You should use `error` in `computeErrors` and call `addErrorMsgField( error )` when both computing methods are finished.

gbirke replied:
used an object instead, see my new commit.

gbirke committed:
Avoid module-global error variable

Created an error storage object instead of using the module-global error
variable. Makes the code much more clear IMO.

gbirke committed:
Fix missing addErrors

gbirke committed:
Check for unknown error messages

Wehn CatGraph sends an unexpected error message, display "Unexpected
result" message to user.

thiemowmde commented on DeepCat.js:
```diff
@@ -81,6 +83,21 @@
+	};
+
+	DeepCat.ResponseErrors.getErrors = function() {
+		return this.errors ? this.errors : [];
```
The JavaScript way of doing this is `return this.errors || []`. :-)

thiemowmde commented on DeepCat.js:
```diff
@@ -81,6 +83,21 @@
 		}
 	} );
 
+	DeepCat.ResponseErrors = {};
```
Please declare and document the `errors` field here.

gbirke committed:
Document DeepCat.ResponseErrors object

gbirke committed:
Make JS more idiomatic

Use OR instead of ternary operator when returning a default value in
getErrors.

WMDE-Fisch commented on DeepCat.js:
```diff
@@ -166,14 +209,15 @@
 			newSearchTerms[userParameters['searchTermNum']] = newSearchTermString;
 		}
 
+		DeepCat.ResponseErrors.addError(errorMessages);
```
Shouldn't this be errorMessages[0] ? You could probably refactor errorMessages here due to the fact, that in this place you only have one case were you create one error message. You could add a conditional clause as well if adding an error is even needed.

gbirke committed:
Fix error message handling

addError is now called with the correct data in computeResponses.
Added explanation why we use an array for collecting the error messages
in computeResponses.

gbirke committed:
Fix the fix

Forgot the index.

WMDE-Fisch commented on DeepCat.js:
```diff
@@ -166,14 +210,17 @@
 		}
 
+		for ( i = 0; i < errorMessages.length; i++ ) {
+			DeepCat.ResponseErrors.addError(errorMessages[i]);
```
Ok last one: Spaces ;-)

gbirke committed:
Fix spaces
